06/merging-sorted-lists.md

```diff
@@ -0,0 +1,214 @@
+
+##### !end-question
+##### !placeholder
+
+```py
+# example input 1:
+# expected output 1:
+
+# example input 2:
+# expected output 2:
+
+def test_nominal_case():
+    # ^rename with meaningful test name
+    # and complete test implementation below
+    pass
+    # arrange
+
+    # act
+
+    # assert
+
+def test_edge_case():
+    # ^rename with meaningful test name
+    # and complete test implementation below
+    pass
+    
+    # arrange
+    
+    # act
+    
+    # assert
+```
+##### !end-placeholder
+
+##### !tests
+
+```py
+import unittest
+
+class TestPython1(unittest.TestCase):
+  def test_always_pass(self):
+    self.assertEqual(1,1)
+```
+
+##### !end-tests
+
+##### !explanation 
+
+Example tests:
+
+```python
+def test_two_nominal_lists():
+    # nominal test case
+    # Arrange
+    list1 = [1, 2, 4, 6]
+    list2 = [5]
+
+    # Act
+    result = merge_lists(list1, list2)
+
+    # Assert
+    assert result == [1, 2, 3, 4, 5, 6]
+
+def test_empty_lists():
+    # edge test case
+    # Arrange
+    list1 = []
+    list2 = []
+
+    # Act
+    result = merge_lists(list1, list2)
+
+    # Assert
+    assert result == []
+
+def test_list_1_all_before_list_2():
+    # alternative test case
+    # Arrange
+    list1 = [-50, -25, 0]
+    list2 = [10, 20, 30]
+
+    # Act
+    result = merge_lists(list1, list2)
+
+    # Assert
+    assert result == [-50, -25, 0, 10, 20, 30]
+
+```
+
+##### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+<!-- Question 3 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: 1f6d0c1e-df66-454b-a6e8-61ce8726b859
+* title: Create Logical Steps
+* points: 3
+* topics: pse
+##### !question
+
+Without writing code, describe how you would implement `merge_lists` in enough detail that someone else could write the code. 
+
+* It may be helpful to break up the problem/algorithm into smaller subproblems/algorithms. For example, 1. Handle invalid input, 2. Given valid input, perform the computation/solve the problem/etc.
+* Your logical steps could take the form of a numbered list, pseudo code, or anywhere in between. What's important at this stage is to think through and outline the implementation before writing code.
+
+##### !end-question
+
+##### !placeholder
+
+Write the logical steps here.
+
+##### !end-placeholder
+
+### !explanation
+
+Example Steps for an O(n<sup>2</sup>) solution:
+
+1. Check for valid input
+1. set i to 0 and j to 0
+1. create result as an empty list
+1. while i < length of list 1 and j < length of list 2
+    - if list[i] < list[j]
+        - append list[i] to result
+        - increment i
+    - else
+        - append list[j] to result
+        - increment j
+1. while i < length of list 1
+    - append list[i] to result
+    - increment i
+1. while j < length of list 2
+    - append list[j] to result
+    - increment j
+1. return result
+
```

Comment:
Minor suggestions to be consistent with previous PSEs and update list to `list1` or `list2`

```suggestion
Example steps for an O(n<sup>2</sup>) solution:

1. Check for valid input
1. set `i` to 0 and `j` to 0
1. initialize `result` as an empty list
1. while `i` < length of `list1` and `j` < length of `list2`
    - if `list1[i]` < `list2[j]`
        - append `list1[i]` to `result`
        - increment `i`
    - else
        - append `list2[j]` to `result`
        - increment `j`
1. while `i` < length of `list1`
    - append `list1[i]` to `result`
    - increment `i`
1. while `j `< length of `list2`
    - append `list2[j]` to `result`
    - increment `j`
1. return `result`

```